torch/csrc/autograd/engine.cpp

```diff
@@ -1117,13 +1117,38 @@ void Engine::add_thread_pool_task(const std::weak_ptr<GraphTask>& graph_task) {
 }
 
 void GraphTask::init_to_execute(Node& graph_root, const edge_list& outputs, bool accumulate_grad) {
+  // Populates exec_info so nodes that should be executed have `exec_info[node].needed_ = true`
+  // Only nodes that have a path to any edge in `outputs` should be executed.
+  // The code below populates exec_info using recursion, but the actual code does this
+  // iteratively. Refer to the numbering to see how the actual code corresponds.
+  //
+  // is_needed = {fn: True for fn in outputs}             # (0)
+  // seen = {}
```

Comment:
shouldn't this be initialized with `outputs`?
Otherwise (1) will never be true for these and you will explore beyond the outputs while it is not needed.

Or if it is needed, that is worth a comment :D

Reply:
Yeah this was a tricky case I ran into. In the case where you have two outputs on the same path, we still want to explore beyond that first output or you would miss all the nodes in between.

Reply:
That's what I was expecting, I think you can add a comment here :)

Reply:
Added!